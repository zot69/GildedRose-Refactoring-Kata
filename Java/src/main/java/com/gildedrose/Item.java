package com.gildedrose;

public class Item {

    public String name;

    public int sellIn;

    public int quality;

    //Made it protected so we use the new Factory
    protected Item(String name, int sellIn, int quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }

    @Override
    public String toString() {
        return this.name + ", " + this.sellIn + ", " + this.quality;
    }


    protected void updateQuality() {
        //default treatment

        if(quality > 0) {
            quality--;
        }

        sellIn--;

        if(sellIn < 0 && quality > 0) {
            quality--;
        }

    }

}
