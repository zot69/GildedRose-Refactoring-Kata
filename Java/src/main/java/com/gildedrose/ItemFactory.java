package com.gildedrose;

import com.gildedrose.itemtypes.Sulfures;
import com.gildedrose.itemtypes.AgedBrie;
import com.gildedrose.itemtypes.ConcertTicket;

public class ItemFactory {
    public static Item create(String name, int sellIn, int quality) {
        switch (name) {
            case "Aged Brie":
                return new AgedBrie(sellIn, quality);
            case "Backstage passes to a TAFKAL80ETC concert":
                return new ConcertTicket(sellIn, quality);
            case "Sulfuras, Hand of Ragnaros":
                return new Sulfures(sellIn, quality);
            default:
                return new Item(name, sellIn, quality);
        }
    }
}
