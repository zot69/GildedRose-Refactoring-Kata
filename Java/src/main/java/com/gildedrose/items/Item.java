package com.gildedrose.items;

public class Item {

    protected String name;
    protected int sellIn;
    protected int quality;

    //Made it protected so we use the new Factory
    protected Item(String name, int sellIn, int quality) {
        this.name = name;
        this.sellIn = sellIn;
        this.quality = quality;
    }

    @Override
    public String toString() {
        return this.name + ", " + this.sellIn + ", " + this.quality;
    }


    public void updateQuality() {
        //default treatment

        if(quality > 0) {
            quality--;
        }

        sellIn--;

        if(sellIn < 0 && quality > 0) {
            quality--;
        }

    }

}
