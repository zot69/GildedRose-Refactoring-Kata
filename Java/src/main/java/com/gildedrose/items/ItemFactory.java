package com.gildedrose.items;

import com.gildedrose.items.types.Conjured;
import com.gildedrose.items.types.Sulfures;
import com.gildedrose.items.types.AgedBrie;
import com.gildedrose.items.types.ConcertTicket;

public class ItemFactory {
    public static Item create(String name, int sellIn, int quality) {
        switch (name) {
            case "Aged Brie":
                return new AgedBrie(sellIn, quality);
            case "Backstage passes to a TAFKAL80ETC concert":
                return new ConcertTicket(sellIn, quality);
            case "Sulfuras, Hand of Ragnaros":
                return new Sulfures(sellIn, quality);
            case "Conjured Mana Cake":
                return new Conjured(sellIn, quality);
            default:
                return new Item(name, sellIn, quality);
        }
    }
}
