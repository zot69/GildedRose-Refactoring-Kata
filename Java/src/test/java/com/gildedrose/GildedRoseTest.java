package com.gildedrose;

import com.gildedrose.items.Item;
import com.gildedrose.items.ItemFactory;
import org.approvaltests.combinations.CombinationApprovals;
import org.junit.Test;

public class GildedRoseTest {

    @Test
    public void updateQuality() throws Exception {

        //Test re-written in order to have a full coverage of the existing code, allowing further refactoring without regressions.
        //I proceed with test coverage tool of the IDE and then by doing some mutation try/error until I get 100% success of this test.

        //Uses CombinationApprovals to avoid writing 50 tests methods and because we can be confident in this code (functional).

        CombinationApprovals.verifyAllCombinations(
                this::executeUpdateQuality,
                new String[]{"Aged Brie", "Backstage passes to a TAFKAL80ETC concert", "Sulfuras, Hand of Ragnaros", "Another item"},
                new Integer[]{0, 11, -1, 6, 3},
                new Integer[]{0, 2, 50, 49});
    }

    private String executeUpdateQuality(String name, int sellIn, int quality) {
        Item[] items = new Item[] { ItemFactory.create(name, sellIn, quality) };
        GildedRose app = new GildedRose(items);
        app.updateQuality();
        return app.getItems()[0].toString();
    }

}
