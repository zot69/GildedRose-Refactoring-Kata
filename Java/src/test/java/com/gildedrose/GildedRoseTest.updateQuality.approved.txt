[Aged Brie, 0, 0] => Aged Brie, -1, 2 
[Aged Brie, 0, 2] => Aged Brie, -1, 4 
[Aged Brie, 0, 50] => Aged Brie, -1, 50 
[Aged Brie, 0, 49] => Aged Brie, -1, 50 
[Aged Brie, 11, 0] => Aged Brie, 10, 1 
[Aged Brie, 11, 2] => Aged Brie, 10, 3 
[Aged Brie, 11, 50] => Aged Brie, 10, 50 
[Aged Brie, 11, 49] => Aged Brie, 10, 50 
[Aged Brie, -1, 0] => Aged Brie, -2, 2 
[Aged Brie, -1, 2] => Aged Brie, -2, 4 
[Aged Brie, -1, 50] => Aged Brie, -2, 50 
[Aged Brie, -1, 49] => Aged Brie, -2, 50 
[Aged Brie, 6, 0] => Aged Brie, 5, 1 
[Aged Brie, 6, 2] => Aged Brie, 5, 3 
[Aged Brie, 6, 50] => Aged Brie, 5, 50 
[Aged Brie, 6, 49] => Aged Brie, 5, 50 
[Aged Brie, 3, 0] => Aged Brie, 2, 1 
[Aged Brie, 3, 2] => Aged Brie, 2, 3 
[Aged Brie, 3, 50] => Aged Brie, 2, 50 
[Aged Brie, 3, 49] => Aged Brie, 2, 50 
[Backstage passes to a TAFKAL80ETC concert, 0, 0] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 2] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 50] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 0, 49] => Backstage passes to a TAFKAL80ETC concert, -1, 0 
[Backstage passes to a TAFKAL80ETC concert, 11, 0] => Backstage passes to a TAFKAL80ETC concert, 10, 1 
[Backstage passes to a TAFKAL80ETC concert, 11, 2] => Backstage passes to a TAFKAL80ETC concert, 10, 3 
[Backstage passes to a TAFKAL80ETC concert, 11, 50] => Backstage passes to a TAFKAL80ETC concert, 10, 50 
[Backstage passes to a TAFKAL80ETC concert, 11, 49] => Backstage passes to a TAFKAL80ETC concert, 10, 50 
[Backstage passes to a TAFKAL80ETC concert, -1, 0] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 2] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 50] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, -1, 49] => Backstage passes to a TAFKAL80ETC concert, -2, 0 
[Backstage passes to a TAFKAL80ETC concert, 6, 0] => Backstage passes to a TAFKAL80ETC concert, 5, 2 
[Backstage passes to a TAFKAL80ETC concert, 6, 2] => Backstage passes to a TAFKAL80ETC concert, 5, 4 
[Backstage passes to a TAFKAL80ETC concert, 6, 50] => Backstage passes to a TAFKAL80ETC concert, 5, 50 
[Backstage passes to a TAFKAL80ETC concert, 6, 49] => Backstage passes to a TAFKAL80ETC concert, 5, 50 
[Backstage passes to a TAFKAL80ETC concert, 3, 0] => Backstage passes to a TAFKAL80ETC concert, 2, 3 
[Backstage passes to a TAFKAL80ETC concert, 3, 2] => Backstage passes to a TAFKAL80ETC concert, 2, 5 
[Backstage passes to a TAFKAL80ETC concert, 3, 50] => Backstage passes to a TAFKAL80ETC concert, 2, 50 
[Backstage passes to a TAFKAL80ETC concert, 3, 49] => Backstage passes to a TAFKAL80ETC concert, 2, 50 
[Sulfuras, Hand of Ragnaros, 0, 0] => Sulfuras, Hand of Ragnaros, 0, 0 
[Sulfuras, Hand of Ragnaros, 0, 2] => Sulfuras, Hand of Ragnaros, 0, 2 
[Sulfuras, Hand of Ragnaros, 0, 50] => Sulfuras, Hand of Ragnaros, 0, 50 
[Sulfuras, Hand of Ragnaros, 0, 49] => Sulfuras, Hand of Ragnaros, 0, 49 
[Sulfuras, Hand of Ragnaros, 11, 0] => Sulfuras, Hand of Ragnaros, 11, 0 
[Sulfuras, Hand of Ragnaros, 11, 2] => Sulfuras, Hand of Ragnaros, 11, 2 
[Sulfuras, Hand of Ragnaros, 11, 50] => Sulfuras, Hand of Ragnaros, 11, 50 
[Sulfuras, Hand of Ragnaros, 11, 49] => Sulfuras, Hand of Ragnaros, 11, 49 
[Sulfuras, Hand of Ragnaros, -1, 0] => Sulfuras, Hand of Ragnaros, -1, 0 
[Sulfuras, Hand of Ragnaros, -1, 2] => Sulfuras, Hand of Ragnaros, -1, 2 
[Sulfuras, Hand of Ragnaros, -1, 50] => Sulfuras, Hand of Ragnaros, -1, 50 
[Sulfuras, Hand of Ragnaros, -1, 49] => Sulfuras, Hand of Ragnaros, -1, 49 
[Sulfuras, Hand of Ragnaros, 6, 0] => Sulfuras, Hand of Ragnaros, 6, 0 
[Sulfuras, Hand of Ragnaros, 6, 2] => Sulfuras, Hand of Ragnaros, 6, 2 
[Sulfuras, Hand of Ragnaros, 6, 50] => Sulfuras, Hand of Ragnaros, 6, 50 
[Sulfuras, Hand of Ragnaros, 6, 49] => Sulfuras, Hand of Ragnaros, 6, 49 
[Sulfuras, Hand of Ragnaros, 3, 0] => Sulfuras, Hand of Ragnaros, 3, 0 
[Sulfuras, Hand of Ragnaros, 3, 2] => Sulfuras, Hand of Ragnaros, 3, 2 
[Sulfuras, Hand of Ragnaros, 3, 50] => Sulfuras, Hand of Ragnaros, 3, 50 
[Sulfuras, Hand of Ragnaros, 3, 49] => Sulfuras, Hand of Ragnaros, 3, 49 
[Another item, 0, 0] => Another item, -1, 0 
[Another item, 0, 2] => Another item, -1, 0 
[Another item, 0, 50] => Another item, -1, 48 
[Another item, 0, 49] => Another item, -1, 47 
[Another item, 11, 0] => Another item, 10, 0 
[Another item, 11, 2] => Another item, 10, 1 
[Another item, 11, 50] => Another item, 10, 49 
[Another item, 11, 49] => Another item, 10, 48 
[Another item, -1, 0] => Another item, -2, 0 
[Another item, -1, 2] => Another item, -2, 0 
[Another item, -1, 50] => Another item, -2, 48 
[Another item, -1, 49] => Another item, -2, 47 
[Another item, 6, 0] => Another item, 5, 0 
[Another item, 6, 2] => Another item, 5, 1 
[Another item, 6, 50] => Another item, 5, 49 
[Another item, 6, 49] => Another item, 5, 48 
[Another item, 3, 0] => Another item, 2, 0 
[Another item, 3, 2] => Another item, 2, 1 
[Another item, 3, 50] => Another item, 2, 49 
[Another item, 3, 49] => Another item, 2, 48 
